File: Classwork/dataStructures/src/mastermind/Mastermind.java
package mastermind;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

import static java.lang.System.out;

public class Mastermind {
	public static void main(String[] args) {
		try (var fileReader = new Scanner(new File("Classwork/dataStructures/src/mastermind/Mastermind.txt"))) {
			var dataSet = new ArrayList<char[]>();
			while (fileReader.hasNextLine())
				dataSet.add(fileReader.nextLine().toCharArray());
			var key = new Object[]{dataSet.get(0), Arrays.toString(dataSet.get(0))};
			for (var i = 1; i < dataSet.size(); i++) {
				var current = dataSet.get(i);
				var matchT1 = 0;
				var failT1 = new ArrayList<Integer>();
				var matchT2 = 0;
				var failT2 = 0;
				for (var j = 0; j < current.length; j++)
					if (((char[]) key[0])[j] == current[j])
						matchT1++;
					else
						failT1.add(j);
				var key2 = new ArrayList<Character>();
				var current2 = new ArrayList<Character>();
				for (int j : failT1) {
					key2.add(((char[]) key[0])[j]);
					current2.add(current[j]);
				}
				for (Character character : current2)
					if (key2.contains(character)) {
						matchT2++;
						key2.remove(character);
					} else
						failT2++;
				out.println("Code:\t" + ((String) key[1]).replaceAll(", ", "").replace("[", "").replace("]", "")
						+ "\nGuess:\t" + Arrays.toString(current).replaceAll(", ", "").replace("[", "").replace("]", "")
						+ "\n\tCorrect Guess - Placement Correct:\t" + matchT1
						+ "\n\tCorrect Guess - Incorrectly Placed:\t" + matchT2 + "\n\tIncorrect Guesses:\t\t\t" + failT2);
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
	}
}

File: Classwork/dataStructures/src/mastermind/Mastermind.kt
package mastermind

import java.io.File
import java.io.FileNotFoundException
import java.util.*

fun main() = try {
	Scanner(File("Classwork/dataStructures/src/mastermind/Mastermind.txt")).use { fileReader ->
		val dataSet = ArrayList<CharArray>()
		while (fileReader.hasNextLine()) dataSet.add(fileReader.nextLine().toCharArray())
		val key = arrayOf(dataSet[0], dataSet[0].contentToString())
		for (i in 1 until dataSet.size) {
			val current = dataSet[i]
			var matchT1 = 0
			val failT1 = ArrayList<Int>()
			var matchT2 = 0
			var failT2 = 0
			current.indices.forEach { if ((key[0] as CharArray)[it] == current[it]) matchT1++ else failT1 += it }
			val key2 = ArrayList<Char>()
			val current2 = ArrayList<Char>()
			for (j in failT1) {
				key2.add((key[0] as CharArray)[j])
				current2.add(current[j])
			}
			current2.forEach {
				if (key2.contains(it)) {
					matchT2++
					key2.remove(it)
				} else failT2++
			}
			println(
				"Code:\t${
					(key[1] as String).replace(", ".toRegex(), "").replace("[", "").replace("]", "")
				}\nGuess:\t${
					current.contentToString().replace(", ".toRegex(), "").replace("[", "").replace("]", "")
				}\n\tCorrect Guess - Placement Correct:\t$matchT1\n\tCorrect Guess - Incorrectly Placed:\t$matchT2\n\tIncorrect Guesses:\t\t\t$failT2"
			)
		}
	}
} catch (e: FileNotFoundException) {
	e.printStackTrace()
}
